base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -3150,41 +3222,48 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
             // lets rewind to last known good accumulated difficulty so the db can be correctly calculated again
             rewind_to_height(db, last_correct_height)?;
         }
-        if migrate_from_version == 2 {
-            // Verify database consistency before starting migration
-            info!(target: LOG_TARGET, "Starting PayRef migration");
+
+        // MIGRATION: Add payref index, rebuild payref index to recover deleted payrefs, add reverse payref index
+        if (migrate_from_version == 2 || migrate_from_version == 3 || migrate_from_version == 4) && !payref_index_done {
+            info!(target: LOG_TARGET, "Starting PayRef migration with reverse lookup");
 
             let read_txn = db.read_transaction()?;
             let chain_height = match fetch_chain_height(&read_txn, &db.metadata_db) {
                 Ok(v) => v,
                 Err(_) => {
                     // No chain data, skip PayRef rebuild
                     drop(read_txn);
-                    continue;
+                    return Ok(());
                 },
             };
             drop(read_txn);
-            // we need to clear the payref index first as they might now bw wrong
+
+            // Clear the payref index first as they might now be wrong
             {
                 let txn = db.write_transaction()?;
                 lmdb_clear(&txn, &db.payref_to_output_index)?;
+                lmdb_clear(&txn, &db.output_to_payref_index)?;
                 txn.commit()?;
                 info!(target: LOG_TARGET, "Cleared PayRef index");
             }
+
             for height in 0..=chain_height {
                 process_payref_for_height(db, height)?;
             }
-            info!(target: LOG_TARGET, "PayRef index rebuild completed");
+
+            payref_index_done = true;
+            info!(target: LOG_TARGET, "PayRef index rebuild with reverse lookup completed");
         }
-        // lets update the migration version
+
+        // Lets update the migration version
         {
             let txn = db.write_transaction()?;
             info!(target: LOG_TARGET, "Migrated database to version {}", MIGRATION_VERSION);
             lmdb_replace(
                 &txn,
                 &db.metadata_db,
                 &k.as_u32(),
-                &MetadataValue::MigrationVersion(migrate_from_version + 1),
+                &MetadataValue::MigrationVersion(max(migrate_from_version + 1, MIGRATION_VERSION)),
```

Comment:
```suggestion
                &MetadataValue::MigrationVersion(max(migrate_from_version , MIGRATION_VERSION)),
```
thats why the migration ended in a 4 and not the correct 3, we can just skip 4 and

Reply:
I solved this in a slightly differtent manner - system-level tested and working